From: Peter Bloomfield <PeterBloomfield@bellsouth.net>
Date: Wed, 22 Jun 2022 12:51:07 -0400
Subject: dataset: Do not increment

Do not increment the `data` pointer when it points to an item that has
not been inspected.

Helps https://gitlab.gnome.org/GNOME/glib/-/issues/2672

Bug: https://gitlab.gnome.org/GNOME/glib/-/issues/2672
Origin: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2776
---
 glib/gdataset.c | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

diff --git a/glib/gdataset.c b/glib/gdataset.c
index 6b78d2e..cc19c96 100644
--- a/glib/gdataset.c
+++ b/glib/gdataset.c
@@ -528,11 +528,8 @@ g_data_remove_internal (GData  **datalist,
               old[found_keys] = *data;
               found_keys++;
 
-              if (data < data_end)
-                {
-                  data_end--;
-                  *data = *data_end;
-                }
+              if (data < --data_end)
+                *data = *data_end;
 
               d->len--;
 
@@ -546,8 +543,10 @@ g_data_remove_internal (GData  **datalist,
                   break;
                 }
             }
-
-          data++;
+          else
+            {
+              data++;
+            }
         }
 
       if (found_keys > 0)
